add defer page to cloud develop docs (#4229)
fishtown-analytics/docs.getdbt.com: website/docs/docs/cloud/about-cloud-develop-defer.md

---
title: Using defer in dbt Cloud
id: about-cloud-develop-defer
description: "Learn how to leverage defer to prod when developing with dbt Cloud."
sidebar_label: "Using defer in dbt Cloud"
pagination_next: "docs/cloud/cloud-cli-installation"
---

> [Defer](/reference/node-selection/defer) is a powerful feature that allows developers to only build and run and test models they've edited without having to first run and build these models' upstream parents. This is powered by having a production manifest to compare to, and dbt will resolve the `{{ ref() }}` function with upstream production artifacts.

Author: ```suggestion
[Defer](/reference/node-selection/defer) is a powerful feature that allows developers to only build and run and test models they've edited without having to first run and build all the models that come before them (upstream parents). This is powered by using a production manifest for comparison, and dbt will resolve the `{{ ref() }}` function with upstream production artifacts.
```

Reviewer: i think i prefer `resolve` but otherwise agree!

Both the dbt Cloud IDE and the dbt Cloud CLI allow users to natively defer to production metadata directly in their development workflows, dramatically reducing development time and warehouse spend by preventing unecessary model builds. 

## Required setup

> Set your production environment using the [production environment checkbox](/docs/deploy/deploy-environments#set-as-production-environment-beta) in the environment settings page. This can be set for one deployment environment per dbt Cloud project. When using defer, the artifacts from the latest successful, non-CI job execution in the production environment will be compared to.

Author: the prod env must be set right? it doesn't work if you have a dev env only?

> Set your production environment using the [production environment checkbox](/docs/deploy/deploy-environments#set-as-production-environment-beta) in the environment settings page. This can be set for one deployment environment per dbt Cloud project. When using defer, the artifacts from the latest successful, non-CI job execution in the production environment will be compared to.

Author: also does the user need to perform a one job run?

Reviewer: yes, there will need to be a successful job run first!

> Set your production environment using the [production environment checkbox](/docs/deploy/deploy-environments#set-as-production-environment-beta) in the environment settings page. This can be set for one deployment environment per dbt Cloud project. When using defer, the artifacts from the latest successful, non-CI job execution in the production environment will be compared to.

Author: ```suggestion
- You must select the **[Production environment](/docs/deploy/deploy-environments#set-as-production-environment-beta)** checkbox in the **Environment Settings** page. 
  - This can be set for one deployment environment per dbt Cloud project. 
  
When using 'defer', it compares artifacts from the most recent successful production job, excluding CI jobs.
```

> ### Defer in the IDE

Author: ```suggestion
### Defer in the dbt Cloud IDE
```

> Defer in the IDE is as simple as toggling the `Defer to production` switch on the command bar! When enabled, dbt Cloud will:

Author: ```suggestion
To enable 'Defer' in the dbt Cloud IDE, toggle the **Defer to production** button on the command bar. Once enabled, dbt Cloud will:
```

Reviewer: typo on `IDEE` -- but great edits

1. Pull down the most recent manifest from the Production environment for comparison

> 2. Pass the `--defer` flag to the command (for any command that accepts the flag)

Author: is this a good example to demonstrate?

```suggestion
2. Pass the `--defer` flag to the command (for any command that accepts the flag). For example, the following command tests `model_b` and uses the most recent manifest from your production environment: 
`dbt test --select model_b --defer --state prod-run-artifacts`
```

Author: also working on clarifying what commands accepts what flags https://github.com/dbt-labs/docs.getdbt.com/issues/4206

Reviewer: that would be a good example but the `--defer` and `--state` flags would be hidden to the user, so not sure we should include

Author: thanks so much for explaining! i'm just a little confused though - how can it be hidden for the user but we're saying they need to pass the `--defer` flag?

Reviewer: it's supposed to indicate that dbt cloud will do the defer on their behalf! they won't need to pass the flag direectly, it will be auto-added in the background

> For example, if you were to start developing with on a new branch with [nothing in your development schema](/reference/node-selection/defer#usage), make an edit to a single model, and run `dbt build -s state:modified` you would see *just* that model run, with any ` {{ ref() }}` functions resolve to the production location of the reffed models!

Author: ```suggestion
For example, if you were to start developing on a new branch with [nothing in your development schema](/reference/node-selection/defer#usage), edit a single model, and run `dbt build -s state:modified` &mdash;  only the edited model would run. Any `{{ ref() }}` functions will point to the production location of the referenced models
```

### Defer in dbt Cloud CLI

> One key difference between using `--defer` in the Cloud CLI vs the IDE is that defer is *automatically on by default for all invocations*, using artifacts from your production evironment to compare. You can pass the `--no-defer` flag to your dbt command to disable this behavior. This flag is unique to the dbt Cloud CLI only.

Author: ```suggestion
One key difference between using `--defer` in the dbt Cloud CLI and the dbt Cloud IDE is that `--defer` is *automatically* enabled in the dbt Cloud CLI for all invocations, comparing with production artifacts. You can disable it with the `--no-defer` flag.
```

Reviewer: my mistake on this one -- i don't believe `--no-defer` is exclusive to the cloud CLI, we can remove that

> Cloud CLI offers additional flexibility for declaring which environment is used as the source of deferral artifacts. You can set a `defer-env-id` key in either your `dbt_project.yml` file or your `dbt_cloud.yml` file.

Author: ```suggestion
The dbt Cloud CLI offers additional flexibility by letting you choose the source environment for deferral artifacts. You can set a `defer-env-id key` in either your `dbt_project.yml` or `dbt_cloud.yml` file.
```

```yml

> # in dbt_cloud.yml

Author: ```suggestion
```

---
title: Using defer in dbt Cloud
id: about-cloud-develop-defer
description: "Learn how to leverage defer to prod when developing with dbt Cloud."
sidebar_label: "Using defer in dbt Cloud"
pagination_next: "docs/cloud/cloud-cli-installation"
---

[Defer](/reference/node-selection/defer) is a powerful feature that allows developers to only build and run and test models they've edited without having to first run and build all the models that come before them (upstream parents). This is powered by using a production manifest for comparison, and dbt will resolve the `{{ ref() }}` function with upstream production artifacts.

Both the dbt Cloud IDE and the dbt Cloud CLI allow users to natively defer to production metadata directly in their development workflows, dramatically reducing development time and warehouse spend by preventing unnecessary model builds. 

## Required setup

- You must select the **[Production environment](/docs/deploy/deploy-environments#set-as-production-environment-beta)** checkbox in the **Environment Settings** page. 
  - This can be set for one deployment environment per dbt Cloud project.

> - You must have a successful job run first 

Author: ```suggestion
- You must have a successful job run first.
```

### Defer in the dbt Cloud IDE

To enable 'Defer' in the dbt Cloud IDE, toggle the **Defer to production** button on the command bar. Once enabled, dbt Cloud will:

1. Pull down the most recent manifest from the Production environment for comparison
2. Pass the `--defer` flag to the command (for any command that accepts the flag)

For example, if you were to start developing on a new branch with [nothing in your development schema](/reference/node-selection/defer#usage), edit a single model, and run `dbt build -s state:modified` &mdash;  only the edited model would run. Any `{{ ref() }}` functions will point to the production location of the referenced models.

<Lightbox src="/img/docs/dbt-cloud/defer-toggle.jpg" width="100%" title="Select the 'Defer to production' toggle on the botom right of the command bar to enable defer in the dbt Cloud IDE."/>

### Defer in dbt Cloud CLI

One key difference between using `--defer` in the dbt Cloud CLI and the dbt Cloud IDE is that `--defer` is *automatically* enabled in the dbt Cloud CLI for all invocations, comparing with production artifacts. You can disable it with the `--no-defer` flag.

> The dbt Cloud CLI offers additional flexibility by letting you choose the source environment for deferral artifacts. You can set a `defer-env-id key` in either your `dbt_project.yml` or `dbt_cloud.yml` file.

Reviewer: ```suggestion
The dbt Cloud CLI offers additional flexibility by letting you choose the source environment for deferral artifacts. You can set a `defer-env-id` key in either your `dbt_project.yml` or `dbt_cloud.yml` file. If you do not provide a `defer-env-id` setting, the dbt Cloud CLI will use artifacts from your dbt Cloud environment marked 'Production'. 
```
